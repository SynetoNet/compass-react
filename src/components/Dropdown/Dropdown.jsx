import React from "react";
import PropTypes from "prop-types";
import Ink from "react-ink";

import BSDropdown from "react-bootstrap/Dropdown";

import "./Dropdown.scss";

class Dropdown extends React.Component {
  static Toggle = BSDropdown.Toggle;
  static Item = BSDropdown.Item;
  static Menu = BSDropdown.Menu;
  static Header = BSDropdown.Header;
  static Divider = BSDropdown.Divider;

  render() {
    const { items, label, variant, children, ...props } = this.props;

    return (
      <BSDropdown {...props}>
        {this.renderTrigger(label, variant)}
        {items ? this.renderItems(items) : children}
      </BSDropdown>
    );
  }

  renderTrigger = (label, variant) => {
    if (typeof label !== "string") {
      const trigger = React.forwardRef((props, ref) => {
        return (
          <div
            {...props}
            ref={ref}
            style={{ cursor: "pointer", display: "inline-block" }}
          >
            {label}
          </div>
        );
      });

      return (
        <Dropdown.Toggle as={trigger} className="dropdown-no-caret">
          {label}
        </Dropdown.Toggle>
      );
    }

    return (
      <Dropdown.Toggle variant={variant}>
        {label}
        <Ink />
      </Dropdown.Toggle>
    );
  };

  renderItems(items) {
    return (
      <Dropdown.Menu>
        {items.map(({ label, eventKey, ...otherProps }) => {
          const key = eventKey || label;
          return (
            <Dropdown.Item key={key} eventKey={key} {...otherProps}>
              {label}
            </Dropdown.Item>
          );
        })}
      </Dropdown.Menu>
    );
  }
}

Dropdown.propTypes = {
  variant: PropTypes.string,
  label: PropTypes.node.isRequired,
  alignRight: PropTypes.bool,
  items: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string.isRequired,
      eventKey: PropTypes.string,
      onSelect: PropTypes.func
    })
  )
};

Dropdown.defaultProps = {
  variant: "secondary",
  alignRight: false
};

export default Dropdown;
