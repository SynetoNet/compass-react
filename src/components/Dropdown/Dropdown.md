#### **Simple Dropdown**

```jsx
<Dropdown
  label="Toggle Dropdown"
  onSelect={eventKey => console.log(eventKey)}
  items={[
    {
      label: "Action",
      eventKey: "action-key",
      onSelect: () => console.log("Additional handler")
    },
    {
      label: "No event key",
      onSelect: () => console.log("eventKey is the label")
    },
    {
      label: "No additional handler"
    }
  ]}
/>
```

#### **Extended usage**

```jsx
<Dropdown label="Toggle Dropdown" onSelect={eventKey => console.log(eventKey)}>
  <Dropdown.Menu>
    <Dropdown.Item eventKey="action">Action</Dropdown.Item>
    <Dropdown.Item eventKey="another">Another action</Dropdown.Item>
    <Dropdown.Item eventKey="other">Something else</Dropdown.Item>
    <Dropdown.Divider />
    <Dropdown.Item onSelect={() => console.log("Separate action")}>
      Separate action
    </Dropdown.Item>
  </Dropdown.Menu>
</Dropdown>
```

#### **Variants**

```jsx
<Dropdown
  label="Primary Dropdown"
  variant="primary"
  onSelect={eventKey => console.log(eventKey)}
  items={[{ label: "Option 1" }, { label: "Option 2" }]}
/>
<div className="mt-3"></div>
<Dropdown
  label="Tertiary Dropdown"
  variant="tertiary"
  onSelect={eventKey => console.log(eventKey)}
  items={[{ label: "Option 1" }, { label: "Option 2" }]}
/>
```
