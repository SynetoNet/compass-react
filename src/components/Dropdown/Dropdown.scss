@import "../../base/colors";
@import "../../base/variables";

.dropdown-toggle::after {
  margin-left: 0.5rem;
  border-top-width: 0.4em;
  transform: translateY(50%);
}

.dropdown-align-end {
  display: flex;
  justify-content: flex-end;
}
