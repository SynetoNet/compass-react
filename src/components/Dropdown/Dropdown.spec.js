import React from "react";
import Dropdown from "./Dropdown";

import { configure, mount, shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

configure({ adapter: new Adapter() });

describe("Default Dropdown", () => {
  const defaultExample = (
    <Dropdown
      show
      label="Dropdown Label"
      items={[{ label: "option1" }, { label: "option2" }]}
    />
  );

  test("renders modal with items prop", () => {
    const component = mount(defaultExample);

    expect(component.contains("Dropdown Label")).toBeTruthy();
    expect(component.contains("option1")).toBeTruthy();
    expect(component.contains("option2")).toBeTruthy();
  });

  test("dropdown with items prop snapshop", () => {
    const component = shallow(defaultExample);
    expect(component).toMatchSnapshot();
  });
});

test("calls onSelect() handlers", () => {
  const selectHandler = jest.fn();

  const example = (
    <Dropdown
      show
      label="Dropdown Label"
      onSelect={selectHandler}
      items={[
        {
          label: "option1",
          eventKey: "option1-key"
        },
        {
          label: "option2",
          eventKey: "option2-key"
        }
      ]}
    />
  );

  const component = mount(example);

  component
    .find(".dropdown-item")
    .at(1)
    .simulate("click");

  component
    .find(".dropdown-item")
    .at(2)
    .simulate("click");

  expect(selectHandler).toHaveBeenNthCalledWith(
    1,
    "option1-key",
    expect.anything()
  );
  expect(selectHandler).toHaveBeenNthCalledWith(
    2,
    "option2-key",
    expect.anything()
  );
  expect(selectHandler).toHaveBeenCalledTimes(2);
});
