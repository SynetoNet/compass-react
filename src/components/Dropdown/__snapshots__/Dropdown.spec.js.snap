// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Default Dropdown dropdown with items prop snapshop 1`] = `
<Dropdown
  navbar={false}
  show={true}
>
  <Bootstrap(DropdownToggle)
    variant="secondary"
  >
    Dropdown Label
    <e
      background={true}
      className="ink"
      duration={1000}
      opacity={0.25}
      radius={150}
      recenter={true}
    />
  </Bootstrap(DropdownToggle)>
  <DropdownMenu
    alignRight={false}
    flip={true}
  >
    <DropdownItem
      as={[Function]}
      disabled={false}
      eventKey="option1"
      key="option1"
    >
      option1
    </DropdownItem>
    <DropdownItem
      as={[Function]}
      disabled={false}
      eventKey="option2"
      key="option2"
    >
      option2
    </DropdownItem>
  </DropdownMenu>
</Dropdown>
`;
